ipam/space/space.go

```diff
@@ -96,26 +103,27 @@ func (s *Space) Free(addr address.Address) error {
 	return nil
 }
 
-func (s *Space) biggestFreeRange() (int, address.Offset) {
+func (s *Space) biggestFreeRange(rangeStart, rangeEnd address.Address) (int, address.Offset) {
 	pos := -1
 	biggest := address.Offset(0)
 
-	for i := 0; i < len(s.free); i += 2 {
-		size := address.Subtract(s.free[i+1], s.free[i])
+	s.walkFree(rangeStart, rangeEnd, func(i int, start, end address.Address) bool {
+		size := address.Subtract(end, start)
 		if size >= biggest {
 			pos = i
 			biggest = size
 		}
-	}
+		return false
+	})
 	return pos, biggest
 }
 
-func (s *Space) Donate() (address.Address, address.Offset, bool) {
+func (s *Space) Donate(rangeStart, rangeEnd address.Address) (address.Address, address.Offset, bool) {
 	if len(s.free) == 0 {
 		return 0, 0, false
 	}
 
-	pos, biggest := s.biggestFreeRange()
+	pos, biggest := s.biggestFreeRange(rangeStart, rangeEnd)
```

Comment:
The code below this looks wrong. We donate a chunk from the end of the biggest free range, but because we are using `pos` this isn't clipped to the requested range.

This would be avoided if `biggestFreeRange` returned the start and end, instead of pos and size.

That would also simplify the new `Space.walkFree`, since we'd no longer need to pass the index to the invoked function.